unit/java_bytecode/inherited_static_fields/inherited_static_fields.cpp

```diff
@@ -0,0 +1,283 @@
+/*******************************************************************\
+
+ Module: Unit tests for inherited static fields
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#include <testing-utils/catch.hpp>
+#include <testing-utils/load_java_class.h>
+
+#include <algorithm>
+
+#include <util/expr_iterator.h>
+
+static bool contains_symbol_reference(const exprt &expr, const irep_idt &id)
+{
+  return
+    std::any_of(
+      expr.depth_begin(),
+      expr.depth_end(),
+      [id](const exprt &e) { // NOLINT (*)
+        return e.id() == ID_symbol && to_symbol_expr(e).get_identifier() == id;
+      });
+}
+
+SCENARIO(
+  "inherited_static_types", "[core][java_bytecode][inherited_static_types]")
+{
+  WHEN("A class uses an inherited static field 'x'")
+  {
+    symbol_tablet symbol_table=
+      load_java_class("Test1", "./java_bytecode/inherited_static_fields");
+    THEN("A static field 'Parent1.x' should exist")
+    {
+      REQUIRE(symbol_table.has_symbol("java::Parent1.x"));
+    }
+    THEN("No static field 'Test1.x' should be created")
+    {
+      REQUIRE(!symbol_table.has_symbol("java::Test1.x"));
+    }
+    THEN("No static field 'java.lang.Object.x' should be created")
+    {
+      REQUIRE(!symbol_table.has_symbol("java::java.lang.Object.x"));
+    }
+    THEN("Test1.main should refer to Parent1.x")
+    {
+      REQUIRE(
+        contains_symbol_reference(
+          symbol_table.lookup_ref("java::Test1.main:()I").value,
+          "java::Parent1.x"));
+    }
+  }
+
+  WHEN("A class uses an interface static field 'x'")
+  {
+    symbol_tablet symbol_table=
+      load_java_class("Test2", "./java_bytecode/inherited_static_fields");
+    THEN("A static field 'StaticInterface2.x' should exist")
+    {
+      REQUIRE(symbol_table.has_symbol("java::StaticInterface2.x"));
+    }
+    THEN("No static field 'Test2.x' should be created")
+    {
+      REQUIRE(!symbol_table.has_symbol("java::Test2.x"));
+    }
+    THEN("No static field 'java.lang.Object.x' should be created")
+    {
+      REQUIRE(!symbol_table.has_symbol("java::java.lang.Object.x"));
+    }
+    THEN("Test2.main should refer to StaticInterface2.x")
+    {
+      REQUIRE(
+        contains_symbol_reference(
+          symbol_table.lookup_ref("java::Test2.main:()I").value,
+          "java::StaticInterface2.x"));
+    }
+  }
+
+  WHEN("A class with an opaque parent uses an inherited static field 'x'")
+  {
+    symbol_tablet symbol_table=
+      load_java_class("Test3", "./java_bytecode/inherited_static_fields");
+    THEN("A static field 'OpaqueParent3.x' should exist")
+    {
```

Comment:
If someone updates Test3.java and recompiles it, the file will come back into existance, could you add a REQUIRE that the symbol for all the opaque classes is correctly marked as an `ID_incomplete_class`?